web/api/v1/api.go

```diff
@@ -1908,9 +1917,19 @@ func parseLimitParam(limitStr string) (limit int, err error) {
 	if err != nil {
 		return limit, err
 	}
-	if limit <= 0 {
-		return limit, errors.New("limit must be positive")
+	if limit < 0 {
+		return limit, errors.New("limit must be non-negative")
 	}
 
 	return limit, nil
 }
+
+// toHintLimit increases the API limit, as returned by parseLimitParam, by 1.
+// This allows for emitting warnings when the results are truncated.
```

Comment:
... when used with storages which do not implement the limit ?

Reply:
This will also be used for storages that implement the limit.
Asking the storage for `user limit+1` let us decide if  with `user limit`, the user sees truncated results.

Reply:
To continue being able to emit these warnings:
https://github.com/prometheus/prometheus/blob/f24ce00320f8605add7c7c5e69f722193c066672/web/api/v1/api.go#L892-L894

Maybe we should make `This allows for emitting warnings when the results are truncated.` clearer.

Reply:
(That being said, we can think about whether we can make the storage itself send such warnings, but maybe in another PR)

Reply:
> ... when used with storages which do not implement the limit ?

After some thought, adding this to the comment seems incorrect. The ONLY reason for this logic is to make the warnings work correctly. If a storage doesn't implement limit, then the warning will work correctly. Only when a storage implements the limit and returns the limit-ed results, the logic for emitting warnings will break.